File: app/src/main/java/pl/mosenko/sunnypodlaskie/persistence/dao/RxBaseDaoImpl.java
package pl.mosenko.sunnypodlaskie.persistence.dao;

import com.j256.ormlite.dao.BaseDaoImpl;
import com.j256.ormlite.dao.CloseableIterator;
import com.j256.ormlite.dao.CloseableWrappedIterable;
import com.j256.ormlite.dao.ForeignCollection;
import com.j256.ormlite.dao.GenericRawResults;
import com.j256.ormlite.dao.ObjectCache;
import com.j256.ormlite.dao.RawRowMapper;
import com.j256.ormlite.dao.RawRowObjectMapper;
import com.j256.ormlite.field.FieldType;
import com.j256.ormlite.stmt.GenericRowMapper;
import com.j256.ormlite.stmt.PreparedDelete;
import com.j256.ormlite.stmt.PreparedQuery;
import com.j256.ormlite.stmt.PreparedUpdate;
import com.j256.ormlite.support.ConnectionSource;
import com.j256.ormlite.support.DatabaseConnection;
import com.j256.ormlite.support.DatabaseResults;
import com.j256.ormlite.table.DatabaseTableConfig;

import java.sql.SQLException;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;

import io.reactivex.Observable;
import pl.mosenko.sunnypodlaskie.persistence.entities.BaseOrmLiteEntity;

/**
 * Created by syk on 17.05.17.
 */

public abstract class RxBaseDaoImpl<DataType extends BaseOrmLiteEntity, IdType> extends BaseDaoImpl<DataType, IdType> implements RxDao<DataType, IdType> {
    protected RxBaseDaoImpl(Class<DataType> dataClass) throws SQLException {
        super(dataClass);
    }

    protected RxBaseDaoImpl(ConnectionSource connectionSource, Class<DataType> dataClass) throws SQLException {
        super(connectionSource, dataClass);
    }

    protected RxBaseDaoImpl(ConnectionSource connectionSource, DatabaseTableConfig<DataType> tableConfig) throws SQLException {
        super(connectionSource, tableConfig);
    }

    @Override
    public Observable<DataType> rxQueryForId(final IdType id) {
        return Observable.fromCallable(() -> queryForId(id));
    }

    @Override
    public Observable<DataType> rxQueryForFirst(final PreparedQuery<DataType> preparedQuery) {
        return Observable.fromCallable(() -> queryForFirst(preparedQuery));
    }

    @Override
    public Observable<List<DataType>> rxQueryForAll() {
        return Observable.fromCallable(() -> queryForAll());
    }

    @Override
    public Observable<List<DataType>> rxQueryForEq(final String fieldName, final Object value) {
        return Observable.fromCallable(() -> queryForEq(fieldName, value));
    }

    @Override
    public Observable<List<DataType>> rxQueryForMatching(final DataType matchObj) {
        return Observable.fromCallable(() -> queryForMatching(matchObj));
    }

    @Override
    public Observable<List<DataType>> rxQueryForMatchingArgs(final DataType matchObj) {
        return Observable.fromCallable(() -> queryForMatchingArgs(matchObj));
    }

    @Override
    public Observable<List<DataType>> rxQueryForFieldValues(final Map<String, Object> fieldValues) {
        return Observable.fromCallable(() -> queryForFieldValues(fieldValues));
    }

    @Override
    public Observable<List<DataType>> rxQueryForFieldValuesArgs(final Map<String, Object> fieldValues) {
        return Observable.fromCallable(() -> queryForFieldValues(fieldValues));
    }

    @Override
    public Observable<DataType> rxQueryForSameId(final DataType data) {
        return Observable.fromCallable(() -> queryForSameId(data));
    }

    @Override
    public Observable<List<DataType>> rxQuery(final PreparedQuery<DataType> preparedQuery) {
        return Observable.fromCallable(() -> query(preparedQuery));
    }

    @Override
    public Observable<Integer> rxCreate(final DataType data) {
        return Observable.fromCallable(() -> create(data));
    }

    @Override
    public Observable<DataType> rxCreateIfNotExists(final DataType data) {
        return Observable.fromCallable(() -> createIfNotExists(data));
    }

    @Override
    public Observable<CreateOrUpdateStatus> rxCreateOrUpdate(final DataType data) {
        return Observable.fromCallable(() -> createOrUpdate(data));
    }

    @Override
    public Observable<Integer> rxUpdate(final DataType data) {
        return Observable.fromCallable(() -> update(data));
    }

    @Override
    public Observable<Integer> rxUpdateId(final DataType data, final IdType newId) {
        return Observable.fromCallable(() -> updateId(data, newId));
    }

    @Override
    public Observable<Integer> rxUpdate(final PreparedUpdate<DataType> preparedUpdate) {
        return Observable.fromCallable(() -> update(preparedUpdate));
    }

    @Override
    public Observable<Integer> rxRefresh(final DataType data) {
        return Observable.fromCallable(() -> refresh(data));
    }

    @Override
    public Observable<Integer> rxDelete(final DataType data) {
        return Observable.fromCallable(() -> delete(data));
    }

    @Override
    public Observable<Integer> rxDeleteById(final IdType id) {
        return Observable.fromCallable(() -> deleteById(id));
    }

    @Override
    public Observable<Integer> rxDelete(final Collection<DataType> datas) {
        return Observable.fromCallable(() -> delete(datas));
    }

    @Override
    public Observable<Integer> rxDeleteIds(final Collection<IdType> ids) {
        return Observable.fromCallable(() -> deleteIds(ids));
    }

    @Override
    public Observable<Integer> rxDelete(final PreparedDelete<DataType> preparedDelete) {
        return Observable.fromCallable(() -> delete(preparedDelete));
    }

    @Override
    public Observable<CloseableIterator<DataType>> rxIterator() {
        return Observable.fromCallable(() -> iterator());
    }

    @Override
    public Observable<CloseableIterator<DataType>> rxIterator(final int resultFlags) {
        return Observable.fromCallable(() -> iterator(resultFlags));
    }

    @Override
    public Observable<CloseableIterator<DataType>> rxIterator(final PreparedQuery<DataType> preparedQuery) {
        return Observable.fromCallable(() -> iterator(preparedQuery));
    }

    @Override
    public Observable<CloseableIterator<DataType>> rxIterator(final PreparedQuery<DataType> preparedQuery, final int resultFlags) {
        return Observable.fromCallable(() -> iterator(preparedQuery, resultFlags));
    }

    @Override
    public Observable<CloseableWrappedIterable<DataType>> rxGetWrappedIterable() {
        return Observable.fromCallable(() -> getWrappedIterable());
    }

    @Override
    public Observable<CloseableWrappedIterable<DataType>> rxGetWrappedIterable(final PreparedQuery<DataType> preparedQuery) {
        return Observable.fromCallable(() -> getWrappedIterable(preparedQuery));
    }

    @Override
    public Observable<GenericRawResults<String[]>> rxQueryRaw(final String query, final String... arguments) {
        return Observable.fromCallable(() -> queryRaw(query, arguments));
    }

    @Override
    public <UO> Observable<GenericRawResults<UO>> rxQueryRaw(final String query, final RawRowMapper<UO> mapper, final String... arguments) {
        return Observable.fromCallable(() -> queryRaw(query, mapper, arguments));
    }

    @Override
    public <UO> Observable<GenericRawResults<UO>> rxQueryRaw(final String query, final com.j256.ormlite.field.DataType[] columnTypes, final RawRowObjectMapper<UO> mapper,
                                                             final String... arguments) {
        return Observable.fromCallable(() -> queryRaw(query, columnTypes, mapper, arguments));
    }

    @Override
    public Observable<GenericRawResults<Object[]>> rxQueryRaw(final String query, final com.j256.ormlite.field.DataType[] columnTypes, final String... arguments) {
        return Observable.fromCallable(() -> queryRaw(query, columnTypes, arguments));
    }

    @Override
    public Observable<Long> rxQueryRawValue(final String query, final String... arguments) {
        return Observable.fromCallable(() -> queryRawValue(query, arguments));
    }

    @Override
    public Observable<Integer> rxExecuteRaw(final String statement, final String... arguments) {
        return Observable.fromCallable(() -> executeRaw(statement, arguments));
    }

    @Override
    public Observable<Integer> rxExecuteRawNoArgs(final String statement) {
        return Observable.fromCallable(() -> executeRawNoArgs(statement));
    }

    @Override
    public Observable<Integer> rxUpdateRaw(final String statement, final String... arguments) {
        return Observable.fromCallable(() -> updateRaw(statement, arguments));
    }

    @Override
    public <CT> Observable<CT> rxCallBatchTasks(final Callable<CT> callable) throws Exception {
        return Observable.fromCallable(() -> callBatchTasks(callable));
    }

    @Override
    public Observable<Boolean> rxObjectsEqual(final DataType data1, final DataType data2) {
        return Observable.fromCallable(() -> objectsEqual(data1, data2));
    }

    @Override
    public Observable<IdType> rxExtractId(final DataType data) {
        return Observable.fromCallable(() -> extractId(data));
    }

    @Override
    public Observable<Class<DataType>> rxGetDataClass() {
        return Observable.fromCallable(() -> getDataClass());
    }

    @Override
    public Observable<FieldType> rxFindForeignFieldType(final Class<?> clazz) {
        return Observable.fromCallable(() -> findForeignFieldType(clazz));
    }

    @Override
    public Observable<Boolean> rxIsUpdatable() {
        return Observable.fromCallable(() -> isUpdatable());
    }

    @Override
    public Observable<Boolean> rxIsTableExists() {
        return Observable.fromCallable(() -> isTableExists());
    }

    @Override
    public Observable<Long> rxCountOf() {
        return Observable.fromCallable(() -> countOf());
    }

    @Override
    public Observable<Long> rxCountOf(final PreparedQuery<DataType> preparedQuery) {
        return Observable.fromCallable(() -> countOf(preparedQuery));
    }

    @Override
    public <FT> Observable<ForeignCollection<FT>> rxGetEmptyForeignCollection(final String fieldName) {
        return Observable.fromCallable(() -> getEmptyForeignCollection(fieldName));
    }

    @Override
    public Observable<ObjectCache> rxGetObjectCache() {
        return Observable.fromCallable(() -> getObjectCache());
    }

    @Override
    public Observable<DataType> rxMapSelectStarRow(final DatabaseResults results) {
        return Observable.fromCallable(() -> mapSelectStarRow(results));
    }

    @Override
    public Observable<GenericRowMapper<DataType>> rxGetSelectStarRowMapper() {
        return Observable.fromCallable(() -> getSelectStarRowMapper());
    }

    @Override
    public Observable<RawRowMapper<DataType>> rxGetRawRowMapper() {
        return Observable.fromCallable(() -> getRawRowMapper());
    }

    @Override
    public Observable<Boolean> rxIdExists(final IdType id) {
        return Observable.fromCallable(() -> idExists(id));
    }

    @Override
    public Observable<DatabaseConnection> rxStartThreadConnection() {
        return Observable.fromCallable(() -> startThreadConnection());
    }

    @Override
    public Observable<Boolean> rxIsAutoCommit(final DatabaseConnection connection) {
        return Observable.fromCallable(() -> isAutoCommit(connection));
    }

    @Override
    public Observable<ConnectionSource> rxGetConnectionSource() {
        return Observable.fromCallable(() -> getConnectionSource());
    }
}

File: app/src/main/java/pl/mosenko/sunnypodlaskie/persistence/dao/RxBaseDaoImpl.kt
package pl.mosenko.sunnypodlaskie.persistence.dao

import com.j256.ormlite.dao.*
import com.j256.ormlite.dao.Dao.CreateOrUpdateStatus
import com.j256.ormlite.field.FieldType
import com.j256.ormlite.stmt.GenericRowMapper
import com.j256.ormlite.stmt.PreparedDelete
import com.j256.ormlite.stmt.PreparedQuery
import com.j256.ormlite.stmt.PreparedUpdate
import com.j256.ormlite.support.ConnectionSource
import com.j256.ormlite.support.DatabaseConnection
import com.j256.ormlite.support.DatabaseResults
import com.j256.ormlite.table.DatabaseTableConfig
import io.reactivex.Observable
import pl.mosenko.sunnypodlaskie.persistence.entities.BaseOrmLiteEntity
import java.util.concurrent.Callable

/**
 * Created by syk on 17.05.17.
 */
abstract class RxBaseDaoImpl<DataType : BaseOrmLiteEntity, IdType> : BaseDaoImpl<DataType, IdType>, RxDao<DataType, IdType> {
    protected constructor(dataClass: Class<DataType>) : super(dataClass)
    protected constructor(connectionSource: ConnectionSource, dataClass: Class<DataType>) : super(connectionSource, dataClass)
    protected constructor(connectionSource: ConnectionSource, tableConfig: DatabaseTableConfig<DataType>) : super(connectionSource, tableConfig)

    override fun rxQueryForId(id: IdType?): Observable<DataType?> {
        return Observable.fromCallable { queryForId(id) }
    }

    override fun rxQueryForFirst(preparedQuery: PreparedQuery<DataType?>?): Observable<DataType?> {
        return Observable.fromCallable { queryForFirst(preparedQuery) }
    }

    override fun rxQueryForAll(): Observable<MutableList<DataType>> {
        return Observable.fromCallable { queryForAll() }
    }

    override fun rxQueryForEq(fieldName: String?, value: Any?): Observable<MutableList<DataType?>?> {
        return Observable.fromCallable { queryForEq(fieldName, value) }
    }

    override fun rxQueryForMatching(matchObj: DataType?): Observable<MutableList<DataType?>?> {
        return Observable.fromCallable { queryForMatching(matchObj) }
    }

    override fun rxQueryForMatchingArgs(matchObj: DataType?): Observable<MutableList<DataType?>?> {
        return Observable.fromCallable { queryForMatchingArgs(matchObj) }
    }

    override fun rxQueryForFieldValues(fieldValues: MutableMap<String?, Any?>?): Observable<MutableList<DataType?>?> {
        return Observable.fromCallable { queryForFieldValues(fieldValues) }
    }

    override fun rxQueryForFieldValuesArgs(fieldValues: MutableMap<String?, Any?>?): Observable<MutableList<DataType?>?> {
        return Observable.fromCallable { queryForFieldValues(fieldValues) }
    }

    override fun rxQueryForSameId(data: DataType?): Observable<DataType?> {
        return Observable.fromCallable { queryForSameId(data) }
    }

    override fun rxQuery(preparedQuery: PreparedQuery<DataType?>?): Observable<MutableList<DataType?>?> {
        return Observable.fromCallable { query(preparedQuery) }
    }

    override fun rxCreate(data: DataType?): Observable<Int?> {
        return Observable.fromCallable { create(data) }
    }

    override fun rxCreateIfNotExists(data: DataType?): Observable<DataType?> {
        return Observable.fromCallable { createIfNotExists(data) }
    }

    override fun rxCreateOrUpdate(data: DataType?): Observable<CreateOrUpdateStatus?> {
        return Observable.fromCallable { createOrUpdate(data) }
    }

    override fun rxUpdate(data: DataType?): Observable<Int?> {
        return Observable.fromCallable { update(data) }
    }

    override fun rxUpdateId(data: DataType?, newId: IdType?): Observable<Int?> {
        return Observable.fromCallable { updateId(data, newId) }
    }

    override fun rxUpdate(preparedUpdate: PreparedUpdate<DataType?>?): Observable<Int?> {
        return Observable.fromCallable { update(preparedUpdate) }
    }

    override fun rxRefresh(data: DataType?): Observable<Int?> {
        return Observable.fromCallable { refresh(data) }
    }

    override fun rxDelete(data: DataType?): Observable<Int?> {
        return Observable.fromCallable { delete(data) }
    }

    override fun rxDeleteById(id: IdType?): Observable<Int?> {
        return Observable.fromCallable { deleteById(id) }
    }

    override fun rxDelete(datas: MutableCollection<DataType?>?): Observable<Int?> {
        return Observable.fromCallable { delete(datas) }
    }

    override fun rxDeleteIds(ids: MutableCollection<IdType?>?): Observable<Int?> {
        return Observable.fromCallable { deleteIds(ids) }
    }

    override fun rxDelete(preparedDelete: PreparedDelete<DataType?>?): Observable<Int?> {
        return Observable.fromCallable { delete(preparedDelete) }
    }

    override fun rxIterator(): Observable<CloseableIterator<DataType?>?> {
        return Observable.fromCallable { iterator() }
    }

    override fun rxIterator(resultFlags: Int): Observable<CloseableIterator<DataType?>?> {
        return Observable.fromCallable { iterator(resultFlags) }
    }

    override fun rxIterator(preparedQuery: PreparedQuery<DataType?>?): Observable<CloseableIterator<DataType?>?> {
        return Observable.fromCallable { iterator(preparedQuery) }
    }

    override fun rxIterator(preparedQuery: PreparedQuery<DataType?>?, resultFlags: Int): Observable<CloseableIterator<DataType?>?> {
        return Observable.fromCallable { iterator(preparedQuery, resultFlags) }
    }

    override fun rxGetWrappedIterable(): Observable<CloseableWrappedIterable<DataType?>?> {
        return Observable.fromCallable { wrappedIterable }
    }

    override fun rxGetWrappedIterable(preparedQuery: PreparedQuery<DataType?>?): Observable<CloseableWrappedIterable<DataType?>?> {
        return Observable.fromCallable { getWrappedIterable(preparedQuery) }
    }

    override fun rxQueryRaw(query: String?, vararg arguments: String?): Observable<GenericRawResults<Array<String?>?>?> {
        return Observable.fromCallable { queryRaw(query, *arguments) }
    }

    override fun <UO> rxQueryRaw(query: String?, mapper: RawRowMapper<UO?>?, vararg arguments: String?): Observable<GenericRawResults<UO?>?> {
        return Observable.fromCallable { queryRaw(query, mapper, *arguments) }
    }

    override fun <UO> rxQueryRaw(query: String?, columnTypes: Array<com.j256.ormlite.field.DataType?>?, mapper: RawRowObjectMapper<UO?>?,
                                 vararg arguments: String?): Observable<GenericRawResults<UO?>?>? {
        return Observable.fromCallable { queryRaw(query, columnTypes, mapper, *arguments) }

    }

    override fun rxQueryRaw(query: String?, columnTypes: Array<com.j256.ormlite.field.DataType?>?, vararg arguments: String?): Observable<GenericRawResults<Array<Any?>?>?> {
        return Observable.fromCallable { queryRaw(query, columnTypes, *arguments) }
    }

    override fun rxQueryRawValue(query: String?, vararg arguments: String?): Observable<Long?> {
        return Observable.fromCallable { queryRawValue(query, *arguments) }
    }

    override fun rxExecuteRaw(statement: String?, vararg arguments: String?): Observable<Int?> {
        return Observable.fromCallable { executeRaw(statement, *arguments) }
    }

    override fun rxExecuteRawNoArgs(statement: String?): Observable<Int?> {
        return Observable.fromCallable { executeRawNoArgs(statement) }
    }

    override fun rxUpdateRaw(statement: String?, vararg arguments: String?): Observable<Int?> {
        return Observable.fromCallable { updateRaw(statement, *arguments) }
    }

    @Throws(Exception::class)
    override fun <CT> rxCallBatchTasks(callable: Callable<CT?>?): Observable<CT?> {
        return Observable.fromCallable { callBatchTasks(callable) }
    }

    override fun rxObjectsEqual(data1: DataType?, data2: DataType?): Observable<Boolean?> {
        return Observable.fromCallable { objectsEqual(data1, data2) }
    }

    override fun rxExtractId(data: DataType?): Observable<IdType?> {
        return Observable.fromCallable { extractId(data) }
    }

    override fun rxGetDataClass(): Observable<Class<DataType?>?> {
        return Observable.fromCallable { getDataClass() }
    }

    override fun rxFindForeignFieldType(clazz: Class<*>?): Observable<FieldType?> {
        return Observable.fromCallable { findForeignFieldType(clazz) }
    }

    override fun rxIsUpdatable(): Observable<Boolean?> {
        return Observable.fromCallable { isUpdatable }
    }

    override fun rxIsTableExists(): Observable<Boolean?> {
        return Observable.fromCallable { isTableExists }
    }

    override fun rxCountOf(): Observable<Long?> {
        return Observable.fromCallable { countOf() }
    }

    override fun rxCountOf(preparedQuery: PreparedQuery<DataType?>?): Observable<Long?> {
        return Observable.fromCallable { countOf(preparedQuery) }
    }

    override fun <FT> rxGetEmptyForeignCollection(fieldName: String?): Observable<ForeignCollection<FT?>?> {
        return Observable.fromCallable { getEmptyForeignCollection(fieldName) }
    }

    override fun rxGetObjectCache(): Observable<ObjectCache?> {
        return Observable.fromCallable { objectCache }
    }

    override fun rxMapSelectStarRow(results: DatabaseResults?): Observable<DataType?> {
        return Observable.fromCallable { mapSelectStarRow(results) }
    }

    override fun rxGetSelectStarRowMapper(): Observable<GenericRowMapper<DataType?>?> {
        return Observable.fromCallable { selectStarRowMapper }
    }

    override fun rxGetRawRowMapper(): Observable<RawRowMapper<DataType?>?> {
        return Observable.fromCallable { rawRowMapper }
    }

    override fun rxIdExists(id: IdType?): Observable<Boolean?> {
        return Observable.fromCallable { idExists(id) }
    }

    override fun rxStartThreadConnection(): Observable<DatabaseConnection?> {
        return Observable.fromCallable { startThreadConnection() }
    }

    override fun rxIsAutoCommit(connection: DatabaseConnection?): Observable<Boolean?> {
        return Observable.fromCallable { isAutoCommit(connection) }
    }

    override fun rxGetConnectionSource(): Observable<ConnectionSource?> {
        return Observable.fromCallable { getConnectionSource() }
    }
}
